### PR TITLE
feat: Code saveing to localStorage on Cmd/Ctrl + S

### PR DESCRIPTION
**Pull request by jakethecoder95**
Had to reopen this PR please see (#37011) for more information.  The flash message right now overlays the editor contents a little bit.  Those changes have been moved to #38327

Right now this is showing a flash message to alert the user that their code was saved.

Before this, the users code was getting saved to localStorage when they attempted to submit the challenge. I've just changed it so that they can manually do it as well. I kept pressing ctrl+s accidently so I thought it would be cool to add this feature. Please let me know if this can be improved in any way.

- [x] I have read [freeCodeCamp's contribution guidelines](https://github.com/freeCodeCamp/freeCodeCamp/blob/master/CONTRIBUTING.md).
- [x] My pull request has a descriptive title (not a vague title like `Update index.md`)
- [x] My pull request targets the `master` branch of freeCodeCamp.
- [x] All the files I changed are in the same world language, for example: only English changes, or only Chinese changes, etc.

### CODE REVIEW

File: client/src/templates/Challenges/redux/code-storage-epic.js [hunk 1]
```diff
@@ -72,17 +74,44 @@ function clearCodeEpic(action$, state$) {
...
+              message: 'Saved! Your code was saved to localStorage'
+            })
+          )
+        : of(
+            createFlashMessage({
+              type: 'warning',
+              message: 'Oops, your code did not save, localStorage may be full'
+            })
+          )
+    )
```
Comment (RandellDawson) [inline client/src/templates/Challenges/redux/code-storage-epic.js]: @jakethecoder95 Any way to make this DRY?  The only difference between these two sections is the `type` and `message`?

Comment (jakethecoder95) [reply client/src/templates/Challenges/redux/code-storage-epic.js]: Sure thing, something like this?
createFlashMessage({
  type: saveSuccessful ? 'success' : 'warning',
  message: saveSuccessful 
    ? 'Saved! Your code was saved to localStorage' 
    : Oops, your code did not save, localStorage may be full'
})

Comment (RandellDawson) [reply client/src/templates/Challenges/redux/code-storage-epic.js]: @jakethecoder95 Something like that.  I don't really like the two ternary statements using `saveSucessful`, but it does seem a bit more concise this way.

**Comment by SanjoSolutions** [comment]
__I would suggest to use the term "local storage" instead of "localStorage" in the messages.__
For me "localStorage" refers to the technical reference `window.localStorage` which the user might not be familiar with. In contrast "local storage" would just refer to some local storage, which still contains the meaning that the code is stored locally in contrast to remotely on a server.
So in my view that would better account for the users which are not familiar with `window.localStorage`. Using camel-case words in an English sentence also seems pretty unusual to me, especially without the technical knowledge of the existence of `window.localStorage`.
For those reasons I find it better to use "local storage" instead of "localStorage in the messages.

**Comment by jakethecoder95** [comment]
> **I would suggest to use the term "local storage" instead of "localStorage" in the messages.**

@SanjoSolutions Thanks for the feedback on this.  I actually wrestled with that question as well when I was writing it.  I stuck with the actual variable name because I thought it was descriptive enough for new users to understand what is happening but also odd looking enough to make them want to look into what it means and do a little research for themselves on localStorage.  Thought it could be a good teaching opportunity. 

Those were just my thoughts.  I am down to change this though! But am interested in seeing if there are more opinions on this first.

**Comment by raisedadead** [comment]
@jakethecoder95 you are correct, we should change the term to local storage, and if I am not mistaken, we should be able to render it as a hyperlink to some additional documentation like for example on MDN

**Comment by RandellDawson** [comment]
> and if I am not mistaken, we should be able to render it as a hyperlink to some additional documentation like for example on MDN

@raisedadead We can, but we made an effort a while back to remove as many external links from the curriculum.  Instead, we tried to link to FCC material on a subject.

**Comment by ojeytonwilliams** [comment]
@raisedadead One thing I noticed while testing this is that `store` will `set` and `get` even if `localStorage` is disabled - it has [several fallbacks](https://www.npmjs.com/package/store#list-of-all-storages).  This could be misleading and imply that something had been persisted when it hadn't.  This is slightly orthogonal to this PR, it's just something we should consider.

@jakethecoder95 the only thing I was not sure about was the mutation of `let saveSuccessful;`.  It seems cleaner to use `map` instead of `tap` and pass the save/fail message through actions, i.e. keep it as functional as possible.  I've got this working locally and l can add it to the PR if you're okay with that.

**Comment by jakethecoder95** [comment]
> **we should be able to render it as a hyperlink to some additional documentation like for example on MDN**

@RandellDawson @raisedadead, before #38062 this was doable but now it looks like the it just shows the anchor tag as a string in the flash message.  Know any way around that?  Also, we could use [freeCodeCamp's localStorage guide](https://guide.freecodecamp.org/javascript/window-localstorage/).  

@ojeytonwilliams Yes please add that to the PR.  I figured there should be a better way to do that but I'm not all that familiar with rxjs yet.

**Comment by RandellDawson** [comment]
> Also, we could use freeCodeCamp's localStorage guide.

@jakethecoder95 No, the guide is deprecated and will eventually cease to exist publicly.  If there is a good news article on https://freecodecamp.org/news` then we could link to it.

**Comment by ojeytonwilliams** [comment]
> I'm not all that familiar with rxjs yet.

Neither am I, honestly.  My reasoning was just that avoiding mutation was a little more 'standard' and hopefully easier to maintain.  Let me know what you think!

**Comment by RandellDawson** [comment]
@ojeytonwilliams With the latest commit using `map` instead of `tap`, is there still a need to import `tap` in `client/src/templates/Challenges/redux/code-storage-epic.js`?

**Comment by ojeytonwilliams** [comment]
@RandellDawson yeah, it's still used by `clearCodeEpic`.  Also, the linter would have complained if it found an unused import.

**Comment by jakethecoder95** [comment]
@ojeytonwilliams Yeah, I think that is much cleaner now. Thank you!  Also, what do you think about adding "dangerouslySetInnerHTML" back into Flash/index.js file so we can add a link?  Is that out of the question?  I see you removed it in January for security reasons.  

@RandellDawson Can't find any freeCodeCamp news articles in there that are specifically about localStorage.  MDN article may be the best bet... or we could leave the link out if external links are not an option.

**Comment by raisedadead** [comment]
Oh yes, I forgot about the not using links in the flash message component. It's too involving to get a link in there. I think its best, we can just call it "your browser's local storage" and that would be it. 

We can surely add a FAQ entry in the https://freecodecamp.org/support for any explanations that we require to do. I do not think we should block the feature in this PR for adding the link which can maybe come later.

**Comment by ojeytonwilliams** [comment]
@jakethecoder95  Pretty much what Mrugesh said.  In principle you could add an extra property to the flash message, `link`, that only creates links.  However, you have to be very careful whenever you don't control the input and make absolutely sure you don't inadvertently open a security hole.

One thing you definitely can't do is `dangerouslySetInnerHTML`!

File: client/src/templates/Challenges/redux/code-storage-epic.js [hunk 2]
```diff
@@ -96,16 +96,17 @@ function saveCodeEpic(action$, state$) {
       }
     }),
     ofType(types.saveEditorContent),
-    switchMap(({ error }) =>
-      of(
+    switchMap(({ error }) => {
+      return of(
```
Comment (ojeytonwilliams) [inline client/src/templates/Challenges/redux/code-storage-epic.js]: Unless I'm missing something, this `return` isn't necessary.  The messages are clearer, though!

Comment (jakethecoder95) [reply client/src/templates/Challenges/redux/code-storage-epic.js]: Oops! forgot to remove that.  I'll fix that now

**Review by ojeytonwilliams — APPROVED** [review]
Nice work @jakethecoder95!  This looks good to me.  I've added a commit, though, so I'd best not merge it.

File: client/src/templates/Challenges/redux/code-storage-epic.js [hunk 3]
```diff
@@ -73,16 +75,38 @@ function clearCodeEpic(action$, state$) {
...
-    ignoreElements()
+    ofType(types.saveEditorContent),
+    switchMap(({ error }) =>
+      of(
+        createFlashMessage({
+          type: error ? 'warning' : 'success',
+          message: error
+            ? // eslint-disable-next-line max-len
+              "Oops, your code did not save, your browser's local storage may be full"
+            : "Saved! Your code was saved to your browser's local storage"
```
Comment (raisedadead) [inline client/src/templates/Challenges/redux/code-storage-epic.js]:
```suggestion
              "Oops, your code did not save, your browser's local storage may be full."
            : "Saved! Your code was saved to your browser's local storage."
```

**Review by raisedadead — CHANGES_REQUESTED** [review]
nitpick

**Comment by raisedadead** [comment]
Hey @ojeytonwilliams, do you want #38327 to land first? Test this locally and go live?

**Review by raisedadead — APPROVED** [review]

**Comment by ojeytonwilliams** [comment]
@raisedadead I'm not fussed about the order. Having the message overlay briefly wouldn't be so bad.

Also, I haven't had chance to look at that PR.

**Comment by raisedadead** [comment]
Thanks for the hard work and patience on this @jakethecoder95, I will push this live in the next release and you should see this live on production in a couple days.

**Pull request merged**